### PR TITLE
Add Sendability annotations

### PR DESCRIPTION
**Pull request by PeterAdams-A**
Motivation:

This becomes a critical feature in swift 6.

Modifications:

Add Sendability annotations where required.
Factor out code protected by locks.

Result:

Compiles without warnings with Sendability checking

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -17,6 +17,7 @@ import PackageDescription
 
 let swiftSettings: [SwiftSetting] = [
     .enableUpcomingFeature("ExistentialAny"),
+    .enableExperimentalFeature("StrictConcurrency=complete"),
```
Comment (Lukasa) [inline Package.swift:20]: I am extremely nervous about enabling this feature here: I'd prefer we did it in CI but left it off in the build.

File: Sources/Logging/LogHandler.swift [hunk 2]
```diff
@@ -169,10 +169,10 @@ public protocol LogHandler: _SwiftLogSendableLogHandler {
     var logLevel: Logger.Level { get set }
 }
 
-extension LogHandler {
+public extension LogHandler {
```
Comment (Lukasa) [inline Sources/Logging/LogHandler.swift:172]: This is the inversion of our preferred style: we prefer to keep the extension undecorated and put visibility on the members. There's a lot of this change in the patch, can it be undone?

File: Sources/Logging/Logging.swift [hunk 3]
```diff
@@ -531,25 +549,28 @@ public enum LoggingSystem {
...
     ///     - metadataProvider: The `MetadataProvider` used to inject runtime-generated metadata from the execution context.
     ///     - factory: A closure that given a `Logger` identifier, produces an instance of the `LogHandler`.
-    public static func bootstrap(_ factory: @escaping (String, Logger.MetadataProvider?) -> any LogHandler,
-                                 metadataProvider: Logger.MetadataProvider?) {
-        self._metadataProviderFactory.replaceMetadataProvider(metadataProvider, validate: true)
-        self._factory.replaceFactory(factory, validate: true)
+    @preconcurrency
+    public static func bootstrap(_ factory: @escaping @Sendable (String, Logger.MetadataProvider?) -> any LogHandler,
+                                 metadataProvider: Logger.MetadataProvider?)
+    {
```
Comment (Lukasa) [inline Sources/Logging/Logging.swift:555]: Can we put this brace back where it was?

File: Sources/Logging/Logging.swift [hunk 4]
```diff
@@ -531,25 +549,28 @@ public enum LoggingSystem {
...
     }
 
     // for our testing we want to allow multiple bootstrapping
-    internal static func bootstrapInternal(_ factory: @escaping (String, Logger.MetadataProvider?) -> any LogHandler,
-                                           metadataProvider: Logger.MetadataProvider?) {
-        self._metadataProviderFactory.replaceMetadataProvider(metadataProvider, validate: false)
-        self._factory.replaceFactory(factory, validate: false)
+    internal static func bootstrapInternal(_ factory: @escaping @Sendable (String, Logger.MetadataProvider?) -> any LogHandler,
+                                           metadataProvider: Logger.MetadataProvider?)
+    {
```
Comment (Lukasa) [inline Sources/Logging/Logging.swift:571]: Same with this brace.

File: Sources/Logging/Logging.swift [hunk 5]
```diff
@@ -580,59 +601,76 @@ public enum LoggingSystem {
...
+        func replace(_ newUnderlying: BoxedType, validate: Bool) {
+            self.storage.withWriteLock { $0.replaceUnderlying(newUnderlying, validate: validate) }
+        }
+
+        var underlying: BoxedType {
+            self.storage.withReadLock { $0.underlying }
         }
     }
+
+    private typealias FactoryBox = ReplaceOnceBox< @Sendable (_ label: String, _ provider: Logger.MetadataProvider?) -> any LogHandler>
```
Comment (Lukasa) [inline Sources/Logging/Logging.swift:666]:
```suggestion
    private typealias FactoryBox = ReplaceOnceBox<@Sendable (_ label: String, _ provider: Logger.MetadataProvider?) -> any LogHandler>
```

File: Sources/Logging/Logging.swift [hunk 6]
```diff
@@ -1107,21 +1159,25 @@ public struct StreamLogHandler: LogHandler {
     internal typealias _SendableTextOutputStream = TextOutputStream & Sendable
 
     /// Factory that makes a `StreamLogHandler` to directs its output to `stdout`
+    @Sendable
```
Comment (Lukasa) [inline Sources/Logging/Logging.swift:1162]: Why is this necessary?

File: Sources/Logging/Logging.swift [hunk 7]
```diff
@@ -1167,7 +1223,8 @@ public struct StreamLogHandler: LogHandler {
                     source: String,
                     file: String,
                     function: String,
-                    line: UInt) {
+                    line: UInt)
+    {
```
Comment (Lukasa) [inline Sources/Logging/Logging.swift:1227]: Return the brace.

**Review by Lukasa — CHANGES_REQUESTED** [review]

File: Sources/Logging/Logging.swift [hunk 8]
```diff
@@ -515,8 +520,9 @@ public enum LoggingSystem {
     ///
     /// - parameters:
     ///     - factory: A closure that given a `Logger` identifier, produces an instance of the `LogHandler`.
-    public static func bootstrap(_ factory: @escaping (String) -> any LogHandler) {
-        self._factory.replaceFactory({ label, _ in
+    @preconcurrency
+    public static func bootstrap(_ factory: @escaping @Sendable(String) -> any LogHandler) {
```
Comment (FranzBusch) [inline Sources/Logging/Logging.swift:524]:
```suggestion
    public static func bootstrap(_ factory: @escaping @Sendable (String) -> any LogHandler) {
```

File: Sources/Logging/Logging.swift [hunk 9]
```diff
@@ -531,25 +537,26 @@ public enum LoggingSystem {
     /// - parameters:
     ///     - metadataProvider: The `MetadataProvider` used to inject runtime-generated metadata from the execution context.
     ///     - factory: A closure that given a `Logger` identifier, produces an instance of the `LogHandler`.
-    public static func bootstrap(_ factory: @escaping (String, Logger.MetadataProvider?) -> any LogHandler,
+    @preconcurrency
+    public static func bootstrap(_ factory: @escaping @Sendable(String, Logger.MetadataProvider?) -> any LogHandler,
```
Comment (FranzBusch) [inline Sources/Logging/Logging.swift:541]:
```suggestion
    public static func bootstrap(_ factory: @escaping @Sendable (String, Logger.MetadataProvider?) -> any LogHandler,
```

File: Sources/Logging/Logging.swift [hunk 10]
```diff
@@ -531,25 +537,26 @@ public enum LoggingSystem {
...
-        self._factory.replaceFactory(factory, validate: true)
+        self._metadataProviderFactory.replace(metadataProvider, validate: true)
+        self._factory.replace(factory, validate: true)
     }
 
     // for our testing we want to allow multiple bootstrapping
-    internal static func bootstrapInternal(_ factory: @escaping (String) -> any LogHandler) {
-        self._metadataProviderFactory.replaceMetadataProvider(nil, validate: false)
-        self._factory.replaceFactory({ label, _ in
+    internal static func bootstrapInternal(_ factory: @escaping @Sendable(String) -> any LogHandler) {
```
Comment (FranzBusch) [inline Sources/Logging/Logging.swift:548]:
```suggestion
    internal static func bootstrapInternal(_ factory: @escaping @Sendable (String) -> any LogHandler) {
```

File: Sources/Logging/Logging.swift [hunk 11]
```diff
@@ -531,25 +537,26 @@ public enum LoggingSystem {
...
+        self._metadataProviderFactory.replace(nil, validate: false)
+        self._factory.replace({ label, _ in
             factory(label)
         }, validate: false)
     }
 
     // for our testing we want to allow multiple bootstrapping
-    internal static func bootstrapInternal(_ factory: @escaping (String, Logger.MetadataProvider?) -> any LogHandler,
+    internal static func bootstrapInternal(_ factory: @escaping @Sendable(String, Logger.MetadataProvider?) -> any LogHandler,
```
Comment (FranzBusch) [inline Sources/Logging/Logging.swift:556]:
```suggestion
    internal static func bootstrapInternal(_ factory: @escaping @Sendable (String, Logger.MetadataProvider?) -> any LogHandler,
```

File: Sources/Logging/Logging.swift (hunk 8 above)
Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift:524]: Fixed

File: Sources/Logging/Logging.swift (hunk 9 above)
Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift:541]: Fixed

File: Sources/Logging/Logging.swift (hunk 11 above)
Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift:556]: Fixed

File: Package.swift (hunk 1 above)
Comment (PeterAdams-A) [reply Package.swift]: Removed.

File: Sources/Logging/Logging.swift (hunk 6 above)
Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift:1147]: Because it's used as a parameter to LoggingSystem.bootstrap which requires it's parameter to be sendable.

File: Sources/Logging/Logging.swift (hunk 5 above)
Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift]: Formatting insists on space.

File: Sources/Logging/Logging.swift (hunk 8 above)
Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift:524]: Formatter insists on lack of space.

File: Sources/Logging/Logging.swift (hunk 11 above)
Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift:556]: Formatter insists on lack of space.

File: Sources/Logging/Logging.swift (hunk 6 above)
Comment (Lukasa) [reply Sources/Logging/Logging.swift:1147]: This is a static function, its parameter _is_ `Sendable`.

Comment (PeterAdams-A) [reply Sources/Logging/Logging.swift]: Ok - this is now removed using the closure trick we discussed offline.

File: docker/docker-compose.yaml [hunk 12]
```diff
@@ -28,7 +28,7 @@ services:
 
   test:
     <<: *common
-    command: /bin/bash -xcl "swift test -Xswiftc -warnings-as-errors --enable-test-discovery $${SANITIZER_ARG-}"
+    command: /bin/bash -xcl "swift test $${WARN_AS_ERROR_ARG-} $${SANITIZER_ARG-} $${IMPORT_CHECK_ARG-} $${EXPLICIT_SENDABLE_ARG-} $${STRICT_CONCURRENCY_ARG}"
```
Comment (Lukasa) [inline docker/docker-compose.yaml:31]:
```suggestion
    command: /bin/bash -xcl "swift test $${WARN_AS_ERROR_ARG-} $${SANITIZER_ARG-} $${IMPORT_CHECK_ARG-} $${EXPLICIT_SENDABLE_ARG-} $${STRICT_CONCURRENCY_ARG-}"
```

**Review by FranzBusch — APPROVED** [review]

**Review by Lukasa — APPROVED** [review]
